Load SSTable at the shard that actually own it

raphaelsc opened the pull request:
Today, the SSTable generation provides a hint on which shard owns a
particular SSTable. That hint determines which shard will load the
SSTable into memory.

With upcoming UUID generation, we will no longer have this hint
embedded into the SSTable generation, meaning that SSTables will be
loaded at random shards. This is not good because shards will have
to reference memory from other shards to access the SSTable
metadata that was allocated elsewhere.

This patch changes sstable_directory to:
1) Use generation value to only determine which shard will calculate
the owner shards for SSTables. Essentially works like a round-robin
distribution.
2) The shard assigned to compute the owners for a SSTable will do
so reading the minimum from disk, usually only Scylla file is
needed.
3) Once that shard finished computing the owners, it will forward
the SSTable to the shard that own it.
4) Shards will later load SSTables locally that were forwarded to
them.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4824/

tchaikov commented on sstables/sstable_directory.hh:
```diff
@@ -53,7 +53,7 @@ class sstable_directory {
     // favor chunked vectors when dealing with file lists: they can grow to hundreds of thousands
     // of elements.
-    using sstable_info_vector = utils::chunked_vector<sstables::foreign_sstable_open_info>;
+    using sstable_open_info_vector = utils::chunked_vector<sstables::foreign_sstable_open_info>;
```
might need to update `replica/distributed_loader.cc` also.

raphaelsc replied:
indeed. one of the reasons for compilation failure. thanks.

avikivity commented on sstables/sstables.cc:
```diff
@@ -1535,6 +1535,36 @@ future<foreign_sstable_open_info> sstable::get_open_info() & {
+                ? scylla_metadata->data.get<scylla_metadata_type::Sharding, sharding_metadata>()
+                : nullptr;
+        return sm && sm->token_ranges.elements.size();
+    }();
```
Tip for the future: using std::invoke(long_lambda) as opposed to long_lambda() is clearer, since it's immediately visible that you're assigning the result of a calculation rather than naming a lambda. With the other syntax, it's only disambiguated at the end.

raphaelsc commented on sstables/sstables.cc:
```diff
@@ -1535,6 +1535,36 @@ future<foreign_sstable_open_info> sstable::get_open_info() & {
+future<std::vector<shard_id>>
+sstable::compute_owner_shards(sstable& sst, const io_priority_class& pc) {
+    co_await sst.read_toc();
+    co_await sst.read_scylla_metadata(pc);
```
TOC content (always less than 4k in size) is required by read_scylla_metadata() to check presence of Scylla component. Note that's also done in other load procedures.

tchaikov replied:
i mean, it's called by `read_scylla_metadata()` already. why shall call it explicitly?

raphaelsc replied:
Ah you're right about that. Will remove explicit read_toc() call in v2. thanks;

raphaelsc commented on sstables/sstables.hh:
```diff
@@ -200,6 +200,10 @@ public:
 
+    // Computes set of shards that own the SSTable.
+    // The method loads only the needed components for computing the set of shards.
+    static future<std::vector<shard_id>> compute_owner_shards(sstable& sst, const io_priority_class& pc = default_priority_class());
```
Will make it non-static in v2.

raphaelsc commented:
v2:
- make compute_owner_shards() non-static and remove uneeded read_toc() call from it.
- fix compilation failure by using new sstable_open_info_vector in distributed_loader.cc

xemul commented:
> With upcoming UUID generation, we will no longer have this hint embedded into the SSTable generation, meaning that SSTables will be loaded at random shards. 

With ownership table we still can have this information. But that's just thinking aloud, it doesn't mean we must do it immediately.

xemul commented on sstables/sstable_directory.cc:
```diff
@@ -109,19 +109,24 @@ void sstable_directory::validate(sstables::shared_sstable sst, process_flags fla
-
+future<sstables::shared_sstable> sstable_directory::load_sstable(sstables::entry_descriptor desc, process_flags flags) const {
     auto sst = _manager.make_sstable(_schema, _sstable_dir.native(), desc.generation, desc.version, desc.format, gc_clock::now(), _error_handler_gen);
     co_await sst->load(_io_priority);
```
Just a random observation that the pattern
```
sst = manager.make_sstable();
co_await sst->load()
return sst
```
is actively used in tests. Maybe it will deserve a helper sugar on manager

raphaelsc replied:
That's true. To be honest, I really hate the fact that sstable became a sink for all operations around SSTable. In my opinion there should be three classes: sstable, sstable_loader, sstable_writer. it's unclear if a sstable is loaded or not. in my opinion, sstable being loaded should be an invariant. sstable_loader loads components for a given sstable and constructs a sstable on completion. Likewise, sstable_writer becomes the sole responsible for writing new sstables and provide a sstable on completion. today, a sstable can be in unloaded, loaded, 'partially written' states. what a mess, no?

tchaikov replied:
does this imply that, with `sstable_loader` , `read_toc()` and `read_scylla_metadata()` will be methods of this class ?

raphaelsc replied:
yes, but that's for the long term. I think the first good step is to split sstable and sstable writer related methods completely. Please note sstable_writer already exists, but we haven't completed the split.

raphaelsc commented:
> > With upcoming UUID generation, we will no longer have this hint embedded into the SSTable generation, meaning that SSTables will be loaded at random shards.
> 
> With ownership table we still can have this information. But that's just thinking aloud, it doesn't mean we must do it immediately.

Indeed. But let's keep in mind the UUID based generation will become the default choice even without object storage.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4846/

tchaikov reviewed (changes requested):
the failed test is actually related to resharding which resumes resharding after increasing the number of shards. i suspect that this change breaks the test.

raphaelsc commented:
> the failed test is actually related to resharding which resumes resharding after increasing the number of shards. i suspect that this change breaks the test.

yes, will fix it.

raphaelsc commented:
v3:
- Rebased
- As it was before, SSTable validation (sstable_directory::validate(...)) is now done in shard designated to process the SSTable. Otherwise, Scylla fails when loading SSTable for materialized view table in a foreign shard. Fixes failure caught by materialized view dtest.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4876/

bhalevy commented on sstables/sstables.cc:
```diff
@@ -1535,6 +1535,35 @@ future<foreign_sstable_open_info> sstable::get_open_info() & {
 
+future<std::vector<shard_id>>
+sstable::compute_owner_shards(const io_priority_class& pc) {
+    co_await read_scylla_metadata(pc);
```
the compute function should just work on a reference to the data structure in memory.
The caller should load it, not the compute function

bhalevy replied:
i.e. we should refactor the logic out of compute_shards_for_this_sstable()

raphaelsc replied:
You have a good point.

Today, if you want to get the shards owning a SSTable, you do;
```
sstable.load();
sstable.get_shards_for_this_sstable();
```

sstable::load() loads the sharding metadata, interpret it and assign result to sstable::_shards.

So it makes sense to have:
`future<> sstable::load_owner_shards();`

which loads the minimum from disk, in order to set sstable::_shards (as sstable::load() does).

And then we can still piggyback on sstable::get_shards_for_this_sstable() to retrieve the list of shards

raphaelsc commented:
v4:
- Rebased
- Split loading from actual retrieval of shards owning the sstable; retrieval is done using existing sstable::get_shards_for_this_sstable().

bhalevy reviewed (approved):
LGTM

raphaelsc commented:
> @scylladb/scylla-maint for some reason, jenkins didn't report the job as success: https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4925/
> 
> please consider merging this PR.

I mean, jenkisn completed the job successfully, but it didn't update GH properly.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4959/
